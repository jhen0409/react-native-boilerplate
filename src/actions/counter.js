export function increment() {
  return {
    type: 'increment',
  };
}

export function decrement() {
  return {
    type: 'decrement',
  };
}

export function incrementIfOdd() {
  return (dispatch, getState) => {
    const { counter } = getState();

    if (counter.get('counter') % 2 === 0) {
      return;
    }

    dispatch(increment());
  };
}

export function incrementAsync(delay = 1000) {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(increment());
    }, delay);
  };
}
