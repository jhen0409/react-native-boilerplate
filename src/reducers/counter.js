import Immutable from 'immutable';

const initialState = Immutable.Map({ counter: 0 });

const actionsMap = {
  increment(state /*, action*/) {
    return state.update('counter', n => n + 1);
  },
  decrement(state /*, action*/) {
    return state.update('counter', n => n - 1);
  },
};

export default (state = initialState, action) => {
  const reduceFn = actionsMap[action.type];
  if (!reduceFn) return state;
  return reduceFn(state, action);
};
