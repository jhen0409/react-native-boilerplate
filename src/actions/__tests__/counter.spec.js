import Immutable from 'immutable';
import * as actions from '../counter';

describe('actions', () => {
  it('increment should create increment action', () => {
    expect(actions.increment()).toMatchSnapshot();
  });

  it('decrement should create decrement action', () => {
    expect(actions.decrement()).toMatchSnapshot();
  });

  it('incrementIfOdd should create increment action', () => {
    const fn = actions.incrementIfOdd();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = jest.fn();
    const getState = () => ({ counter: Immutable.Map({ counter: 1 }) });
    fn(dispatch, getState);
    expect(dispatch).toHaveBeenCalledWith({ type: 'increment' });
  });

  it('incrementIfOdd shouldnt create increment action if counter is even', () => {
    const fn = actions.incrementIfOdd();
    const dispatch = jest.fn();
    const getState = () => ({ counter: Immutable.Map({ counter: 2 }) });
    fn(dispatch, getState);
    expect(dispatch).not.toHaveBeenCalled();
  });

  // There's no nice way to test this at the moment...
  it('incrementAsync', (done) => {
    const fn = actions.incrementAsync(1);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = jest.fn();
    fn(dispatch);
    setTimeout(() => {
      expect(dispatch).toHaveBeenCalledWith({ type: 'increment' });
      done();
    }, 5);
  });
});
