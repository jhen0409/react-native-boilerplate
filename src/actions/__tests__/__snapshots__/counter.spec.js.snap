// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions decrement should create decrement action 1`] = `
Object {
  "type": "decrement",
}
`;

exports[`actions increment should create increment action 1`] = `
Object {
  "type": "increment",
}
`;
